morpheus/_lib/include/morpheus/messages/control.hpp

```diff
@@ -0,0 +1,160 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <nlohmann/json.hpp>
+#include <pybind11/pybind11.h>
+
+#include <memory>
+
+namespace morpheus {
+class MessageMeta;
+#pragma GCC visibility push(default)
+enum class ControlMessageType
+{
+    INFERENCE,
+    NONE,
+    TRAINING
+};
+
+class MessageControl
+{
+  public:
+    MessageControl();
+    MessageControl(const nlohmann::json& config);
+    MessageControl(const MessageControl& other);  // NO payload copy
+
+    /**
+     * @brief Set the config object
+     * @param config
+     */
+    void config(const nlohmann::json& config);
+
+    /**
+     * @brief Get the config object
+     * @return
+     */
+    const nlohmann::json& config() const;
+
+    /**
+     * @brief Add a task to the control message
+     * @param task
+     * @param type
+     */
+    void add_task(const std::string& task_type, const nlohmann::json& task);
+
+    /**
+     * @brief Check if a task of a given type exists
+     * @param type
+     * @return
+     */
+    bool has_task(const std::string& task_type) const;
+
+    /**
+     * @brief Get a task of the given type
+     * @param type
+     * @return
+     */
+    const nlohmann::json pop_task(const std::string& task_type);
+
+    /**
+     * @brief Add a metadata key-value pair to the control message
+     * @param key
+     * @param value
+     */
+    void set_metadata(const std::string& key, const nlohmann::json& value);
+
+    /**
+     * @brief Check if a metadata key exists
+     * @param key
+     * @return
+     */
+    bool has_metadata(const std::string& key) const;
+
+    // TODO(Devin): Allow for a default value
+    /**
+     * @brief Get the metadata value for a given key
+     * @param key
+     * @return
+     */
+    const nlohmann::json get_metadata(const std::string& key) const;
+
+    /**
+     * @brief Set the payload object
+     * @param payload
+     */
+    void payload(const std::shared_ptr<MessageMeta>& payload);
+
+    /**
+     * @brief Get the payload object
+     * @return Shared pointer to the message payload
+     */
+    std::shared_ptr<MessageMeta> payload();
+
+    /**
+     * @brief Get the type of the task
+     * @return ControlMessageType
+     */
+    ControlMessageType task_type();
+
+    /**
+     * @brief Set the task type for the control message
+     * @param task_type
+     * @return
+     */
+    void task_type(ControlMessageType task_type);
+
+  private:
+    static const std::string s_config_schema;  // NOLINT
+
+    ControlMessageType m_cm_type{ControlMessageType::NONE};
+    std::shared_ptr<MessageMeta> m_payload{nullptr};
+    std::map<std::string, ControlMessageType> m_task_type_map{{"inference", ControlMessageType::INFERENCE},
+                                                              {"training", ControlMessageType::TRAINING}};
+
+    nlohmann::json m_tasks{};
```

Comment:
Currently, no. I'd thought about having a priority, or some type of sorting metric, but I didn't have any required use case, and it could be added later.